### PR TITLE
Add support for passthrough with directory remapping

### PR DESCRIPTION
**Pull request by MadeByMike**
Modifies the `addPassthroughCopy` function so that it accepts a 2nd param that allows you to remap the output directory.

This means you can do something like the following:
```
config.addPassthroughCopy({"src/site/static", "/"});
```
This will map all files and folders to the root directory. 

After reading [this issue](https://github.com/11ty/eleventy/issues/214)  I've updated it so that you can use globs. This will map all js files to the root directory:

```
config.addPassthroughCopy({"src/site/static/*.js": "/"});
```
Because of the way that globs work, it's not possible to retain nested directory structures and all matched files will be output to the specified directory. For example, this will find all file in nested folders named `js` and will output them to the 'js` folder in the root:
```
config.addPassthroughCopy({"src/site/static/**/js/*": "/js"});
```
This means if there are multiple `index.js` files in these folders they will be overwritten by the last copy.

To get around this and keep the option to recursively copy directories if the inputPath is not a glob and matches a directory, the directory will be recursively copied to the output path. For example, here all files in `src/static/js` will be copied to the `js` folder in root including any all sub directories:
```
config.addPassthroughCopy({"src/site/static/js": "/js"});
```

For backward compatibility passing `true` also works, and the output would be resolved to `static/js`:
```
config.addPassthroughCopy({"src/site/static/js": true});
```
Not passing an object has the same result:
```
config.addPassthroughCopy("src/site/static/js");
```

### CODE REVIEW

File: src/TemplatePassthrough.js [hunk 1]
```diff
@@ -3,35 +3,33 @@ const TemplatePath = require("./TemplatePath");
...
     this.isDryRun = false;
   }
 
   getOutputPath() {
-    return TemplatePath.join(
-      this.outputDir,
-      TemplatePath.stripLeadingSubPath(this.path, this.inputDir)
-    );
+    const { inputDir, outputDir, inputPath, outputPath } = this;
+    // assuming if paths are the same an outputPath was not set and we will resolve manually?
```
Comment (MadeByMike) [inline src/TemplatePassthrough.js]: See comment in PR. I don't love method of testing if the output path was set. If it's not set we resolve an output path that includes the last directory, if it is set we don't.

That means in theory these to examples _should_ have different output paths:
```
config.addPassthroughCopy("./src/site/static"); // output path: "/static"
```
```
config.addPassthroughCopy("./src/site/static", "./src/site/static"); // output path SHOULD be: "/src/site/static" however will equal "/static"
```
```
config.addPassthroughCopy("./src/site/static", "./src/site/static/deeper"); // output path: "/src/site/static/deeper"
```
You can see that adding one more directory in this **edge case** would result in a significant difference to the output path. It's an edge case and it's understandable. However If the code can be improved it probably should. (want help\feedback)

File: src/TemplatePassthrough.js [hunk 2]
```diff
@@ -3,35 +3,33 @@ const TemplatePath = require("./TemplatePath");
...
+        : outputPath;
+    return TemplatePath.join(outputDir, path);
   }
 
   setDryRun(isDryRun) {
     this.isDryRun = !!isDryRun;
   }
 
   async write() {
-    // debug(
```
Comment (MadeByMike) [inline src/TemplatePassthrough.js]: clean up :)

File: src/TemplatePassthroughManager.js [hunk 3]
```diff
@@ -8,9 +8,8 @@ const debug = require("debug")("Eleventy:TemplatePassthroughManager");
 class TemplatePassthroughManagerCopyError extends EleventyBaseError {}
 
 class TemplatePassthroughManager {
-  constructor(inputDir, outputDir, isDryRun) {
```
Comment (MadeByMike) [inline src/TemplatePassthroughManager.js]: not used

File: src/TemplatePassthroughManager.js [hunk 4]
```diff
@@ -45,15 +44,21 @@ class TemplatePassthroughManager {
     let target = this.config.passthroughCopies || {};
     debug("`passthroughFileCopy` config paths: %o", target);
     for (let path in target) {
-      paths.push(TemplatePath.addLeadingDotSlash(path));
+      const inputPath = TemplatePath.addLeadingDotSlash(path);
+      const outputPath =
+        typeof target[path] === "string"
```
Comment (MadeByMike) [inline src/TemplatePassthroughManager.js]: If string, assume an explicit output path. Else assume boolean and match input path

**Comment by edwardhorsford** [comment]
@MadeByMike [this issue](https://github.com/11ty/eleventy/issues/214) is possibly relevant to you.

**Comment by MadeByMike** [comment]
Thanks @edwardhorsford it's pretty close to what is suggested there maybe we can build on this.

**Comment by julrichkieffer** [comment]
👍  can't wait for this enhancement!

**Comment by MadeByMike** [comment]
Tracked down a bug now which is why CI failed....

If you have: 

config.addPassthroughCopy({
  "src/js/*.js": "/all/js", 
  "src/static": "/all"
});

The order of execution matters! recursive file copying will attempt to make a directory that already exists

**Comment by kleinfreund** [comment]
Can you add tests for paths that escape the input or output directory, please?

```js
config.addPassthroughCopy({
  "static": "..", // escapes output directory
  "..": "." // escapes input directory
});
```

They should throw an error.

Also, I’d prefer if no examples or tests use a leading forward slash as it usually represents the file system’s root directory.

And finally, what does the following do? I don’t understand why there would be a boolean.

```js
config.addPassthroughCopy({
  "static": true,
});
```

**Comment by MadeByMike** [comment]
> I don’t understand why there would be a boolean

@kleinfreund It's how the old API worked. I would not expect anyone to do this, but that's what happens behind the scenes when you do `config.addPassthroughCopy("static")`. I needed to make sure it still works.

**Comment by jgarber623** [comment]
@MadeByMike Thank you so much for implementing this feature!

Quick question regarding:

> **Because of the way that globs work**, it's not possible to retain nested directory structures and all matched files will be output to the specified directory.

(emphasis mine)

Could you explain that a bit more? Is this a path resolution issue with the way Node/JavaScript handles globbing?

For reference, I'm coming from the Ruby world where code like `Dir.glob('./src/_assets/stylesheets/**/*.scss')` would return an unsorted array with matched paths prefixed with the original query preceding the globbed characters (e.g. `./src/_assets/stylesheets/`):

```
[
  "./src/_assets/stylesheets/application.scss", 
  "./src/_assets/stylesheets/components/_page-header.scss", 
  "./src/_assets/stylesheets/components/_page-footer.scss", 
  "./src/_assets/stylesheets/components/_buttons.scss", 
  "./src/_assets/stylesheets/components/_page-navigation.scss", 
  "./src/_assets/stylesheets/components/_hero.scss", 
  "./src/_assets/stylesheets/base/_variables.scss", 
  "./src/_assets/stylesheets/base/_utilities.scss", 
  "./src/_assets/stylesheets/base/_elements.scss"
]
```

**Comment by MadeByMike** [comment]
@jgarber623 yes exactly right! So imagine our key was the glob you mentioned and the output folder was `./assets`: 
```
{ './src/_assets/stylesheets/**/*.scss': './assets' }
```
We need to make a decision about what this means because it is ambiguous and what we choose will impose some limitations on users. My assumption is to match all .scss files and move them to the `./assets` folder. (This is what has been implemented)

However you could also assume that the user intends to copy all .scss files from `./src/_assets/stylesheets/` and then copy these to './assets' keeping the folders in the `stylesheets` directory. This is easy for a human to understand but not so easy to find a clear set of rule that work with all globs.

If we match globs and then map files to a single directory. This means you only get the last file, if the match returns 2 with the same name. Eg:

```
// If the glob result was:
[
"./src/_assets/stylesheets/base/index.scss"
"./src/_assets/stylesheets/typography/index.scss"
] 
// The output is:
[
"./assets/index.css" // From typography
] 
```

If it's not a glob we can recursively copy the directory.

```
// Result is:
[
"./src/_assets/stylesheets/"
] 
// The output is:
[
  "./src/_assets/stylesheets/application.scss", 
  "./src/_assets/stylesheets/components/_page-header.scss", 
  "./src/_assets/stylesheets/components/_page-footer.scss", 
  "./src/_assets/stylesheets/components/_buttons.scss", 
  "./src/_assets/stylesheets/components/_page-navigation.scss", 
  "./src/_assets/stylesheets/components/_hero.scss", 
  "./src/_assets/stylesheets/base/_variables.scss", 
  "./src/_assets/stylesheets/base/_utilities.scss", 
  "./src/_assets/stylesheets/base/_elements.scss"
]
// Along with any other non .scss files!
```
So it's that 3rd option of being able to recursively copy a directory and filter the results by regex that is tricky. I suggest that it should not be in scope for this PR but maybe we make a new issue to resolve it immediately after merging this?

**Comment by MadeByMike** [comment]
@kleinfreund I wrote a test and updated code to catch files attempting to be written outside the site output directory but I can figure out how to silence the errors in the console. If you'd like to help with this I'd love feedback.

@zachleat I'd love your input on this one too!

**Comment by MadeByMike** [comment]
I'm kinda of the opinion that anything that returns a promise should always return a promise and you reject rather than throw. This gives a better call stack in the error messages and makes it easier to catch on `Promise.all()`.

I dunno I'm sure there are people with better opinions on this and who can articulate it better.  

I may be wrong on that

**Comment by kleinfreund** [comment]
I’m not sure why the errors are still thrown. Something is broken.

The example test in [ava Assertions: `.throwsAsync(thrower, [expected, [message]])`](https://github.com/avajs/ava/blob/master/docs/03-assertions.md#throwsasyncthrower-expected-message) copied into the tests doesn’t spill its error out like that so that’s something on our end. The following works as expected:

```js
test('throws', async t => {
	await t.throwsAsync(async () => {
		throw new TypeError('🦄');
	}, {instanceOf: TypeError, message: '🦄'});
});
```

Another note: Some of your tests have duplicate entries in the `passthroughCopies` object (e.g. `"../": true` is overwritten by `"../": "./"`). Please replace one of these instances with a different input path so both are actually tested.

And a question: What issue is commit https://github.com/11ty/eleventy/pull/452/commits/96fbd65edca6e292303b2d58e03b486d0013a2cf fixing? I don’t think we should normalize Windows path separators anywhere in the code.

File: src/TemplatePassthrough.js [hunk 5]
```diff
@@ -1,41 +1,98 @@
...
+        this.copy(inputFile, this.getGlobOutputPath(inputFile), copyOptions)
+      );
 
-      return copy(this.path, this.getOutputPath(), {
-        overwrite: true,
-        dot: true,
-        junk: false,
-        results: false
+      return Promise.all(promises).catch(err => {
+        throw new EleventyBaseError(
```
[the remark this answers is not in the document]
Comment (MadeByMike) [reply src/TemplatePassthrough.js]: Sure :) - Thanks

File: src/TemplatePassthrough.js [hunk 6]
```diff
@@ -1,41 +1,98 @@
...
+      const promises = files.map(inputFile =>
+        this.copy(inputFile, this.getGlobOutputPath(inputFile), copyOptions)
+      );
 
-      return copy(this.path, this.getOutputPath(), {
-        overwrite: true,
-        dot: true,
-        junk: false,
-        results: false
+      return Promise.all(promises).catch(err => {
```
[the remark this answers is not in the document]
Comment (MadeByMike) [reply src/TemplatePassthrough.js]: Yes, and to be honest I'm not sure what the best way to handle it is hence my comments [here](#issuecomment-491664603) and [here](#issuecomment-491664052). @zachleat any chance you can give me some implementation advice here?

File: src/TemplatePassthrough.js [hunk 7]
```diff
@@ -1,41 +1,98 @@
...
+
+    if (outputPath === true) {
+      return TemplatePath.join(
+        outputDir,
+        TemplatePath.stripLeadingSubPath(inputPath, inputDir)
+      );
+    }
+    return path
+      .normalize(TemplatePath.join(outputDir, outputPath))
+      .replace(/\\/g, "/"); // Fix for windows paths
```
[the remark this answers is not in the document]
Comment (MadeByMike) [reply src/TemplatePassthrough.js]: Still fails but I think I can remove it if I path.normalize() all the paths in tests. Hopefully we're joining paths correctly everywhere else.

File: src/TemplatePassthrough.js [hunk 8]
```diff
@@ -1,41 +1,98 @@
...
+  }
+
+  getGlobOutputPath(globFile) {
```
[the remark this answers is not in the document]
Comment (MadeByMike) [reply src/TemplatePassthrough.js]: Maybe `getGlobOutputPath` is a bad name. The glob files are first resolved [here](https://github.com/MadeByMike/eleventy/blob/master/src/TemplatePassthrough.js#L84). Then the relative file path (not a glob string) is passed to `getGlobOutputPath` this gets the last path segment and effectively 'flattens' the globed files into the specified output directory.

[the remark this answers is not in the document]
Comment (MadeByMike) [reply src/TemplatePassthrough.js]: I renamed this to `getOutputPathForGlobFile` for more clarity. This is tested indirectly in many of the `TemplatePassthrough` tests and directly in the "Glob pattern" test.

File: src/TemplatePassthrough.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (MadeByMike) [reply src/TemplatePassthrough.js]: Took me ages to track down! It was the difference between using `TemplatePath.normalize` vs native `path.normalize`. Not sure why we're using `TemplatePath.normalize` ? But it's consistent now and I didn't want to change any more existing code.

File: src/TemplatePassthrough.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (zachleat) [reply src/TemplatePassthrough.js]: After a bit more review it looks like this case only hits when you’re using glob paths right? It keeps the previous behavior with files or directory. I think that’s okay.

**Comment by zachleat** [comment]
Hey! I fixed the conflicts here and am going to merge this but as a follow up I am opening an issue for the last remaining piece of feedback here. #615

**Comment by MadeByMike** [comment]
OMG OMG MERGED 😂😂😂

**Comment by zachleat** [comment]
Shipping with the next version!
